AP_Scripting: add bindings for servo telemetry

IamPete1 opened the pull request:
Adds bindings for servo telem, this allows servo checks to be done in lua.

IamPete1 commented on libraries/AP_Scripting/docs/docs.lua:
```diff
@@ -4013,3 +4012,58 @@ function visual_odom:healthy() end
+
+
+-- servo telemetry class
+---@class servo_telem
```
Don't need this.

IamPete1 commented:
For refence attempt 1 cost 1500 bytes or so. 
![image](https://github.com/user-attachments/assets/f4192021-76c7-4a1b-be35-f882fe70aa1c)

IamPete1 commented:
This now get the structure and adds auto generation of a bitmask check and set. This means the script writer does not need to know there is a validity bitmask in the structure. This is currently read only, but it also adds auto setting of the mask when a value is set. The only slight problem is that once set there is no way to un-set the mask from lua. 

One disadvantage to this method is that were stuck with the types in the structure, previously I had returned the temp in degrees by applying the scale in the getter, now were stuck with centi degrees.

IamPete1 commented on libraries/AP_Scripting/docs/docs.lua:
```diff
@@ -1395,9 +1395,6 @@ function camera:take_picture(instance) end
 ---@class (exact) AP_Camera__camera_state_t_ud
 local AP_Camera__camera_state_t_ud = {}
 
----@return AP_Camera__camera_state_t_ud
-function AP_Camera__camera_state_t() end
-
```
This should not have been in the docs generation, the function is not generated on the C++ side. We do not need it because all fields are read only and there are no methods or operators. The docs generator has been fixed to match the C++ generator.

tpwrules replied:
I found another tiny similar bug where we were accidentally generating `lua_new_X` functions that were never called. I folded that fix into this commit.

IamPete1 replied:
@tpwrules I disagree with your changes. We can emit them and rely on the linker to not include them if there not used. It seems it does still build, but you might want to use one of the functions in a manual binding which the generator does not know about. What is the benefit of not generating them?

tpwrules replied:
It was my understanding that those functions were specifically designed to be called by Lua, e.g. because of the warnings and the use of the stack. If a manual binding wanted to create a new object, it would use the `new_X` function.

You are right that there is no grave harm in generating them though but I don't think a manual binding using them makes a lot of sense.

tpwrules replied:
In case it wasn't clear here are the two places I added a new `should_emit_creation()` call to cause the change in behavior:
* https://github.com/ArduPilot/ardupilot/blob/e72720f2bffd5abda7908e927032f1adfcf707b2/libraries/AP_Scripting/generator/src/main.c#L1346
* https://github.com/ArduPilot/ardupilot/blob/e72720f2bffd5abda7908e927032f1adfcf707b2/libraries/AP_Scripting/generator/src/main.c#L1431

IamPete1 replied:
Your right, we don't need those.

IamPete1 commented on libraries/AP_Scripting/docs/docs.lua:
```diff
@@ -4011,7 +4008,6 @@ function networking:get_netmask_active() end
 function networking:get_ip_active() end
 
 -- visual odometry object
---@class visual_odom
```
This is incorrect markup. It should be `---@class`, however the class mark up is for user data objects only and should not be used on singletons. The error was overlooked due to the typo.

IamPete1 commented:
Example of the new docs markup:
```
userdata AP_Servo_Telem::TelemetryData field command_position float'skip_check read valid_mask valid_types AP_Servo_Telem::TelemetryData::Types::COMMANDED_POSITION
```
First bit is the same, the stuff added is:
```
valid_mask valid_types AP_Servo_Telem::TelemetryData::Types::COMMANDED_POSITION
```
"valid_mask" is the keyword to enable the bitmask validity check.
"valid_types" is the variable name on the "AP_Servo_Telem::TelemetryData" object. 
"AP_Servo_Telem::TelemetryData::Types::COMMANDED_POSITION" is the bitmask value that the mask should be checked against.

IamPete1 commented:
Now only costs  about 1K, so saves 400 bytes over the individual getters. 

![image](https://github.com/user-attachments/assets/e917dae3-c0f3-4dba-9d26-f0e08257b042)

tpwrules commented:
I haven't looked in detail, but it seems this can't absolutely fix the race condition because there isn't a lock protecting the telemetry data. The race window is much shorter now however, so maybe it's fixed enough? It was said in the call that there's servos which output different data at different rates anyway, so maybe comparing commanded vs. achieved just isn't possible in general.

Also still curious what the size impact would be if you had one mega function which returned eleven values.

rmackay9 reviewed (approved):
Looks OK to me but @tpwrules will review more thoroughly

tpwrules reviewed (changes requested):
A couple little things in the generator but should be ready to go after!

I tested in SITL using quadplane-can and the position from four servos comes through but it's the same for all four of them. I checked that the data I could get from Lua matched the data logged in CSRV, and it all did so this PR is good, but there's still probably something funky here.

IamPete1 commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -1041,3 +1041,23 @@ include AP_TemperatureSensor/AP_TemperatureSensor.h
+
+userdata AP_Servo_Telem::TelemetryData depends AP_SERVO_TELEM_ENABLED
+userdata AP_Servo_Telem::TelemetryData rename AP_Servo_Telem_Data
+userdata AP_Servo_Telem::TelemetryData field command_position float'skip_check read valid_mask valid_types AP_Servo_Telem::TelemetryData::Types::COMMANDED_POSITION
```
Yeah, they should all be skip. But it doesn't matter in this case because there read only. Otherwise I think we should have got a error. 

I will fix.

IamPete1 replied:
fixed.

IamPete1 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -2578,20 +2606,14 @@ void emit_loaders(void) {
-  fprintf(source, "} new_userdata[] = {\n");
-  while (data) {
+int should_emit_creation(struct userdata * data) {
     // Dont expose creation function for all read only items
```
done.

IamPete1 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -2601,8 +2623,18 @@ void emit_userdata_new_funcs(void) {
+  fprintf(source, "    const lua_CFunction fun;\n");
+  fprintf(source, "} new_userdata[] = {\n");
+  while (data) {
+    // Dont expose creation function for all read only items
```
removed

IamPete1 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -2885,32 +2917,33 @@ void emit_docs(struct userdata *node, int is_userdata, int emit_creation) {
-        if (node->creation != NULL) {
-          for (int i = 0; i < node->creation_args; ++i) {
-            fprintf(docs, "---@param param%i UNKNOWN\n", i+1);
+      if (should_emit_creation(node)) {
```
Yeah, I have re-worked to simplify.

IamPete1 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -2922,11 +2955,14 @@ void emit_docs(struct userdata *node, int is_userdata, int emit_creation) {
     if (node->fields != NULL) {
       struct userdata_field *field = node->fields;
       while(field) {
+          int valid_mask = (field->type.valid_mask.name != NULL) & (field->type.valid_mask.value != NULL);
```
done

IamPete1 commented on libraries/AP_Scripting/generator/src/main.c:
```diff
@@ -2922,11 +2955,14 @@ void emit_docs(struct userdata *node, int is_userdata, int emit_creation) {
       struct userdata_field *field = node->fields;
       while(field) {
+          int valid_mask = (field->type.valid_mask.name != NULL) & (field->type.valid_mask.value != NULL);
+          const char * return_postfix = valid_mask ? "|nil\n" : "*\n";
```
typo.

tpwrules reviewed (approved):
Tested and still works, thanks for the fixes.

I cleaned up the history slightly and folded in another really tiny fix to the generator. Figured it was easier to do it myself then to go through review again.

tpwrules commented:
Concerned whether servo telemetry needs a fix similar to https://github.com/ArduPilot/ardupilot/pull/28999 but that doesn't impact this PR at all.

tridge commented on libraries/AP_Servo_Telem/AP_Servo_Telem.cpp:
```diff
@@ -152,6 +152,38 @@ void AP_Servo_Telem::write_log()
+
+    const volatile TelemetryData &telem_data = _telem_data[servo_index];
+
+    // Because the structure is volatile we have to copy element wise
```
we could just cast to override the volatile?

tridge replied:
or could memcpy?

IamPete1 replied:
I couldn't get casting to work. The compiler was too smart. `memcpy` may work....

tridge replied:
try this: 
```
    telem = *const_cast<TelemetryData*>(&telem_data);
```

tpwrules replied:
`memcpy(&telem, (const void*)&telem_data, sizeof(telem));` will work too. But it's not exactly kosher.

IamPete1 replied:
> try this:
> 
> ```
>     telem = *const_cast<TelemetryData*>(&telem_data);
> ```

I have changed to this. Thanks.

robertlong13 commented:
> Concerned whether servo telemetry needs a fix similar to #28999 but that doesn't impact this PR at all.

AP_Servo_Telem works subtly differently and does not suffer the same issue. The getter functions return success if that telemetry field has _ever_ been reported, and it is up to the caller to check whether that data is stale. AP_ESC_Telem's getters return failure if the data has timed out (or if a race condition leads it to falsely conclude there has been a timeout).

`AP_Servo_Telem::stale` suffers the same potential race condition, but that function is only used by `AP_Servo_Telem::valid` and _that_ function is unused. The lua bindings use the `valid_mask` instead of `AP_Servo_Telem::valid`, which should probably be named to `present_mask` since it's the mask checked by `AP_Servo_Telem::present` and `AP_Servo_Telem::valid` requires things to be both present and fresh.

Perhaps that rename should be part of this PR though.

IamPete1 commented:
> `AP_Servo_Telem::stale` suffers the same potential race condition, but that function is only used by `AP_Servo_Telem::valid` and _that_ function is unused. The lua bindings use the `valid_mask` instead of `AP_Servo_Telem::valid`, which should probably be named to `present_mask` since it's the mask checked by `AP_Servo_Telem::present`. `AP_Servo_Telem::valid` requires things to be both present and fresh.

Yep, your right. I have removed the unused functions and renamed `valid_types` to `present_types`
